### ISSUE TITLE
Make the library header only

### ISSUE THREAD

**Issue by DraTeots** [SRombauts/SQLiteCpp#132]
Is there anything, that prevents converting this library into a header only library? 

I believe it would be very convenient in many aspects.

**Comment by SRombauts** [comment]
Yes, sure, this should be very convenient. I went for this at the very beginning, but then wanted to move to more traditional dev and reduce compile times inherent to headers only.

If one can make a Python script to generate a big header file out of the many sources files, I could upload the monolithic header with each release (somewhat like the sqlite Amalgation source file)

**Comment by DraTeots** [comment]
I could look into it. I saw something like this in the [Catch unit test framework](https://github.com/philsquared/Catch). While their script combines only headers, maybe It would be easy to adjust it.

**Comment by DraTeots** [comment]
Looked the code, python scripts and thinking about the particular python script. I would prefer not to use full scale C++ parsing facilities. 

So would it be acceptable make some current code styling to be required. Like putting comments about closing name spaces? 

```c++
}  // namespace SQLite
```
Such comments are already there but only as a matter of better readability, while they will become required for the script to work. Same goes for some other formatting choices like putting `extern const` definitions in one line. 

```c++
extern const .... <VeriableName>;    // GOOD

extern const                   // BAD
.... <VeriableName>;    
```

**Comment by SRombauts** [comment]
Yes sure, also such namespace comments are already enforced by the Python cpplint script :)

**Comment by DraTeots** [comment]
Ok! Suppose I have a python file that generates a single header (I actually do):

```
single_header/SQLiteCpp.h
```
It compiles, tests are passed. How to integrate it in CMake build, tests, CI etc? Need your opinion as a library author:

1. Is ```single_header/SQLiteCpp.h``` OK? 

    Or `single_header/SQLiteCpp/SQLiteCpp.h`?

    Better naming? Ideas? 

2. What I did now, I copied (**bad!**) tests to a separate directory and changed their includes to a single header version. What I would do is to introduce one more definition option like  ```TEST_SINGLE_HEADER``` and make all tests like:

    ```c++
    #ifndef TEST_SINGLE_HEADER
    #include <SQLiteCpp/Backup.h>
    #include <SQLiteCpp/Database.h>
    #else
    #include <SQLiteCpp.h>
    #endif   // TEST_SINGLE_HEADER
    ```
    ```TEST_SINGLE_HEADER```?  
     ```TEST_HEADER_ONLY```? - better naming?

3. Add  one more target with ```TEST_SINGLE_HEADER``` definition to CMake. Something like this:

    ```cmake
       add_executable(SQLiteCpp_ho_tests ${SQLITECPP_TESTS})
       target_compile_definitions(SQLiteCpp_sh_tests TEST_SINGLE_HEADER)
    ```
    And use it in CI
    Or is there a better way? 

4. One has to add ```-std=c++11``` to compile with the single header. 

5. This single include definitely shell not pass cpplint. I think this is OK.

**Comment by SRombauts** [comment]
Hi @DraTeots, I am sorry but I am very busy at the moment...

We should probably move one step by one step: if you can provide the scrip, I can run it manually to release the monolithic header as a downloadable file.

I am not too eager to complexify the CMake files nor do I need to automate testing on it.

Cheers!

**Comment by pratikpc** [comment]
@SRombauts check out [amalgamate](https://github.com/matthewcpp/cpp-amalgamate)

**Comment by SRombauts** [comment]
Closing this now due to age and inactivity; please open a fresh request if a single-header distribution is still wanted.
